src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexMatcher.Automata.cs

```diff
@@ -0,0 +1,429 @@
+            lock (this)
+            {
+                // Get the next state if it exists.  The caller should have already tried and found it null (not yet created),
+                // but in the interim another thread could have created it.
+                (int, DerivativeEffect[])[]? targets = _capturingNfaDelta[offset];
+                if (targets is null)
+                {
+                    DfaMatchingState<TSet> coreState = GetState(GetCoreStateId(nfaStateId));
+                    TSet minterm = GetMintermFromId(mintermId);
+                    uint nextCharKind = GetPositionKind(mintermId);
+                    List<(SymbolicRegexNode<TSet> Node, DerivativeEffect[] Effects)>? transition = coreState.NfaNextWithEffects(_builder, minterm, nextCharKind);
+                    // Build the new state and store it into the array.
+                    List<(int, DerivativeEffect[])> targetsList = new();
+                    foreach ((SymbolicRegexNode<TSet> Node, DerivativeEffect[] Effects) entry in transition)
+                    {
+                        ForEachNfaState(entry.Node, nextCharKind, (targetsList, entry.Effects),
+                            static (int nfaId, (List<(int, DerivativeEffect[])> Targets, DerivativeEffect[] Effects) args) =>
+                                args.Targets.Add((nfaId, args.Effects)));
+                    }
+                    targets = targetsList.ToArray();
+                    Volatile.Write(ref _capturingNfaDelta[offset], targets);
+                }
+
+                return targets;
+            }
+        }
+
+        /// <summary>
+        /// Iterates through the alternation branches <see cref="SymbolicRegexNode{TSet}.EnumerateAlternationBranches(SymbolicRegexBuilder{TSet})"/>
+        /// and tries to create NFA states for each. The supplied action is called for each created NFA state. These never
+        /// include dead ends as <see cref="CreateNfaState(SymbolicRegexNode{TSet}, uint)"/> will filter those out.
+        /// </summary>
+        /// <remarks>This function locks the matcher for safe concurrent use of the <see cref="_builder"/></remarks>
+        /// <typeparam name="T">the type of the additional argument passed through to the action</typeparam>
+        /// <param name="node">the node to break up into NFA states</param>
+        /// <param name="prevCharKind">the previous character kind for each created NFA state</param>
+        /// <param name="arg">an additional argument passed through to each call to the action</param>
+        /// <param name="action">action to call for each NFA state</param>
+        private void ForEachNfaState<T>(SymbolicRegexNode<TSet> node, uint prevCharKind, T arg, Action<int, T> action)
+        {
+            lock (this)
+            {
+                foreach (SymbolicRegexNode<TSet> nfaNode in node.EnumerateAlternationBranches(_builder))
+                {
+                    if (CreateNfaState(nfaNode, prevCharKind) is int nfaId)
```

Comment:
It's ok that this is forcing the state into existence / caching it?

Reply:
I think it's fine, although not perfect. The only states that aren't given an NFA ID for are ones that simplify to "nothing" after the anchor pruning is done, so there's a low limit to how many times that can happen (once for each prevCharKind).